Firestore/Example/Tests/Integration/API/FIRValidationTests.m

```diff
@@ -161,13 +161,66 @@ - (void)testWritesWithNonDictionaryValuesFail {
   }
 }
 
-- (void)testWritesWithNestedArraysFail {
+- (void)testWritesWithDirectlyNestedArraysFail {
   [self expectWrite:@{
     @"nested-array" : @[ @1, @[ @2 ] ]
   }
       toFailWithReason:@"Nested arrays are not supported"];
 }
 
+- (void)testWritesWithIndirectlyNestedArraysSucceed {
+  NSDictionary<NSString *, id> *data = @{ @"nested-array" : @[ @1, @{@"foo" : @[ @2 ]} ] };
+
+  FIRDocumentReference *ref = [self documentRef];
+  FIRDocumentReference *ref2 = [self documentRef];
+
+  XCTestExpectation *expectation = [self expectationWithDescription:@"setData"];
+  [ref setData:data
+      completion:^(NSError *_Nullable error) {
+        XCTAssertNil(error);
+        [expectation fulfill];
+      }];
+  [self awaitExpectations];
+
+  expectation = [self expectationWithDescription:@"batch.setData"];
+  [[[ref.firestore batch] setData:data forDocument:ref]
+      commitWithCompletion:^(NSError *_Nullable error) {
+        XCTAssertNil(error);
+        [expectation fulfill];
+      }];
+  [self awaitExpectations];
+
+  expectation = [self expectationWithDescription:@"updateData"];
+  [ref updateData:data
+       completion:^(NSError *_Nullable error) {
+         XCTAssertNil(error);
+         [expectation fulfill];
+       }];
+  [self awaitExpectations];
+
+  expectation = [self expectationWithDescription:@"batch.updateData"];
+  [[[ref.firestore batch] updateData:data forDocument:ref]
+      commitWithCompletion:^(NSError *_Nullable error) {
+        XCTAssertNil(error);
+        [expectation fulfill];
+      }];
+  [self awaitExpectations];
+
+  XCTestExpectation *transactionDone = [self expectationWithDescription:@"transaction done"];
+  [ref.firestore runTransactionWithBlock:^id(FIRTransaction *transaction, NSError **pError) {
+    // Note ref2 does not exist at this point so set that and update ref.
+    [transaction updateData:data forDocument:ref];
+    [transaction setData:data forDocument:ref2];
+    return nil;
+  }
+      completion:^(id result, NSError *error) {
+        // ends up being a no-op transaction.
+        XCTAssertNil(error);
+        [transactionDone fulfill];
+      }];
+  [self awaitExpectations];
```

Comment:
FWIW, I commend you for hitting all the write paths, but in the name of expediency / simplicity, I wouldn't fault you for just covering one (or maybe setData / updateData both) and adding a comment something like "// Technically we should perhaps test all the write APIs, but they hit the same validation code path so omitted for brevity."  especially when you port this twice.

Reply:
I only did this on all the paths because the negative cases are also hitting all the paths.

I think I'd prefer this test to be a unit test verifying all the permutations and then an integration test verifying that validation is performed at all on each of the write paths. That's not a change I want to do here though (if you even agree).

Reply:
Yeah...  The negative cases use a single helper giving us a nice ROI though.  FWIW, this whole file is actually only the negative cases (part of my logic was to put them in one place so we could easily sanity-check all the error messages).  

So perhaps you should just move this test to FIRDatabaseTests.m or FIRTypeTests.m and then get rid of the exhaustive coverage since you don't need to be consistent with FIRValidationTests. :p

Reply:
I've left this in here for now: as much as this is a positive test it's a test of the validation logic.